crates/walrus-core/src/encoding/quilt_encoding.rs

```diff
@@ -0,0 +1,1367 @@
+    fn encode(&self) -> Result<Vec<SliverPair>, QuiltError>;
+
+    /// Encodes the blobs into a quilt and returns the slivers and metadata.
+    fn encode_with_metadata(&self) -> Result<(Vec<SliverPair>, V::QuiltMetadata), QuiltError>;
+}
+
+pub trait QuiltDecoderApi<'a, V: QuiltVersion> {
+    /// Decodes the quilt index from received slivers.
+    ///
+    /// The decoded quilt index is stored in the decoder and can be retrieved
+    /// using the `get_quilt_index` method after this method returns.
+    fn decode_quilt_index(&mut self) -> Result<&V::QuiltIndex, QuiltError>;
+
+    /// Returns the decoded quilt index, if available.
+    fn get_quilt_index(&self) -> Option<&V::QuiltIndex>;
+
+    /// Gets a blob by its identifier from the quilt.
+    fn get_blob_by_identifier(&self, identifier: &str) -> Result<Vec<u8>, QuiltError>;
+
+    /// Adds slivers to the decoder.
+    fn add_slivers(&mut self, slivers: &'a [&'a SliverData<Secondary>]);
+}
+
+pub trait QuiltApi<V: QuiltVersion> {
+    /// Gets a blob by its identifier from the quilt.
+    fn get_blob_by_identifier(&self, identifier: &str) -> Result<Vec<u8>, QuiltError>;
+
+    /// Returns the quilt index.
+    fn quilt_index(&self) -> &V::QuiltIndex;
+
+    /// Returns the data of the quilt.
+    fn data(&self) -> &[u8];
+
+    /// Returns the symbol size of the quilt.
+    fn symbol_size(&self) -> usize;
+}
+
+pub struct QuiltVersionV1;
+
+impl QuiltVersion for QuiltVersionV1 {
+    type QuiltConfig = QuiltConfigV1;
+    type QuiltEncoder<'a> = QuiltEncoderV1<'a>;
+    type QuiltDecoder<'a> = QuiltDecoderV1<'a>;
+    type Quilt = QuiltV1;
+    type QuiltIndex = QuiltIndexV1;
+    type QuiltMetadata = QuiltMetadata;
+}
+
+/// The configuration of a quilt.
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+pub struct QuiltConfigV1 {}
+
+impl<'a> QuiltConfigApi<'a, QuiltVersionV1> for QuiltConfigV1 {
+    fn get_encoder(
+        &self,
+        encoding_config: EncodingConfigEnum<'a>,
+        blobs: &'a [BlobWithIdentifier<'a>],
+    ) -> QuiltEncoderV1<'a> {
+        QuiltEncoderV1::new(encoding_config, blobs)
+    }
+
+    fn get_decoder(&self, slivers: &'a [&'a SliverData<Secondary>]) -> QuiltDecoderV1<'a> {
+        QuiltDecoderV1::new(slivers)
+    }
+
+    fn new_from_quilt_blob(
+        &self,
+        quilt_blob: Vec<u8>,
+        metadata: &QuiltMetadata,
+        n_shards: NonZeroU16,
+    ) -> Result<QuiltV1, QuiltError> {
+        // assert_eq!(self, &QuiltConfig::V1);
+        let encoding_config = EncodingConfig::new(n_shards);
+        let config = encoding_config.get_for_type(metadata.metadata().encoding_type());
+
+        let n_primary_source_symbols = config.n_primary_source_symbols().get();
+        let n_secondary_source_symbols = config.n_secondary_source_symbols().get();
+        let n_source_symbols = n_primary_source_symbols * n_secondary_source_symbols;
+
+        // Verify data alignment.
+        if quilt_blob.len() % n_source_symbols as usize != 0 {
```

Comment:
General remark: Try to replace ` as ` conversions by something like `from()`/`into()` or `try_from()/try_into()` to make it clear that/why the conversion is correct.